### PR TITLE
feat: add floating window

### PR DESCRIPTION
**Pull request by KKuze**
you can start-up foating window to keydown ctrl-h in '*.super'

### CODE REVIEW

**Comment by asana17** [comment]
I think you should create appropriate help message instead of usikg temporary messages such as `aaaaa`.

Please fix it, I will review this PR after the fix.

File: lua/super-commit/window/float.lua [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        "hint3",
+        "hint4"
+    }
+
+    vim.api.nvim_buf_set_lines(buf, 0, -1, false, messages)
+
+    local screen_width = vim.o.columns
+    local screen_height = vim.o.lines
+
+    local msglen = 0;
```
Comment (asana17) [inline lua/super-commit/window/float.lua:45]:
```suggestion
    local max_len = 0;
```

File: lua/super-commit/window/float.lua [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    for _, msg in pairs(messages) do
+        msglen = (msglen > string.len(msg)) and msglen or string.len(msg)
+    end
+    msglen = msglen <= 80 and msglen or 80
+    local width = screen_width >= 160 and math.floor(screen_width / 2) or msglen
```
Comment (asana17) [inline lua/super-commit/window/float.lua:50]: Here you are updating the width based on the editor size, despite the previous implementation calculates the max length of help strings.

Probably the previous calculation of strings are unnecessary.

**Review by asana17 — CHANGES_REQUESTED** [review]

**Comment by asana17** [comment]
Also, the current implementation is good for showing help floating window, but it is difficult to use the implementation for other floating windows.

File: lua/super-commit/window/float.lua [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    vim.api.nvim_buf_set_lines(buf, 0, -1, false, messages)
+
+    local screen_width = vim.o.columns
+    local screen_height = vim.o.lines
+
+    local msglen = 0;
+    for _, msg in pairs(messages) do
+        msglen = (msglen > string.len(msg)) and msglen or string.len(msg)
+    end
+    msglen = msglen <= 80 and msglen or 80
```
Comment (manhatsu) [inline lua/super-commit/window/float.lua:49]: It's better to define fixed values (such as `80` or `160`) as local variables if you are going to use them multiple times.

File: lua/super-commit/window/float.lua (hunk 1 above)
Comment (KKuze) [reply lua/super-commit/window/float.lua:45]: Fixed it.

File: lua/super-commit/window/float.lua (hunk 2 above)
Comment (KKuze) [reply lua/super-commit/window/float.lua]: Fixed as `local max_width = math.floor(screen_width * 2 / 3)`.

File: lua/super-commit/window/float.lua (hunk 3 above)
Comment (KKuze) [reply lua/super-commit/window/float.lua]: The numbers are deleted for the above fix.

File: lua/super-commit/window/float.lua [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    vim.api.nvim_buf_set_lines(buf, 0, -1, false, messages)
+
+    local screen_width = vim.o.columns
+    local screen_height = vim.o.lines
+
+    local max_width = math.floor(screen_width * 2 / 3)
+    local width = screen_width >= 160
+                and math.floor(screen_width / 2)
+                or max_width
```
Comment (asana17) [inline lua/super-commit/window/float.lua:19]: For the maximum length, fixed 80 characters are enough for the readability. Thinking of some margin, set a larger fixed size. Don't forget to check whether it is centralized.

Also for the small editor witdh, I suggest to use its full length.

The actual problem occurs when the editor size is small. This implementation still shrinks the floating window sized based on editor width.

Suppose the window editor width equals to "40". Don't you think utilizing the whole window is better than using 2/3?

File: lua/super-commit/window/float.lua [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    local height = math.min(
+            math.floor(screen_height / 2),
+            #messages
+        )
+
+    local opts = {
+      relative = "editor",
+      width = width,
+      height = height,
+      col = (screen_width - width) / 2,
```
Comment (asana17) [inline lua/super-commit/window/float.lua:29]: Is floor division unnecessary here?

**Review by asana17 — CHANGES_REQUESTED** [review]

File: lua/super-commit/window/float.lua (hunk 3 above)
Comment (asana17) [reply lua/super-commit/window/float.lua]: I think `160` still remains in the code. These magic numbers are better to defined as another variable to make the code clearer.

See this [Stack Overflow Q&A](https://stackoverflow.com/questions/47882/what-are-magic-numbers-and-why-do-some-consider-them-bad)

File: lua/super-commit/window/float.lua (hunk 4 above)
Comment (manhatsu) [reply lua/super-commit/window/float.lua:19]: I agree with this.
And considering the meaning, use `min_width` instead of `max_width`.

Comment (manhatsu) [reply lua/super-commit/window/float.lua:19]:
```suggestion
    local max_width = 80 -- or any
    local width = math.min(screen_width, max_width)
```

File: .luacheckrc [hunk 6]
```diff
@@ -11,5 +11,6 @@ globals = {
   "require",
   "vim",
   "pairs",
-  "math"
+  "math",
+  "string"
```
Comment (asana17) [inline .luacheckrc:15]: `string` is not used anymore.

**Review by asana17 — CHANGES_REQUESTED** [review]

File: .luacheckrc (hunk 6 above)
Comment (KKuze) [reply .luacheckrc]: Fixed it in 219b40b43c2596f5984d95db27a7910f2b0ff9c2.

File: lua/super-commit/window/float.lua (hunk 2 above)
Comment (KKuze) [reply lua/super-commit/window/float.lua]: Fixed it.

File: lua/super-commit/window/float.lua (hunk 3 above)
Comment (KKuze) [reply lua/super-commit/window/float.lua]: Fixed it.

File: lua/super-commit/window/float.lua [hunk 7]
```diff
@@ -0,0 +1,41 @@
+local M = {}
+
+function M.help()
+    local buf = vim.api.nvim_create_buf(false, true)
+    local messages = {
+        "You can write",
+        "a message of",
+        "this app informations."
+    }
```
[the remark this answers is not in the document]
Comment (KKuze) [reply lua/super-commit/window/float.lua]: Fixed it.

File: lua/super-commit/window/float.lua (hunk 4 above)
Comment (KKuze) [reply lua/super-commit/window/float.lua]: Fixed It as const value.

File: lua/super-commit/window/float.lua (hunk 5 above)
Comment (KKuze) [reply lua/super-commit/window/float.lua]: Fixed It as const value.

**Pull request merged**